### PR TITLE
Re-land typed descriptors and printer

### PR DESCRIPTION
**Pull request by sebmarkbage**
Release notes: landing two previously reverted PRs

I found the issue.

Flow doesn't support optional fields in classes. This effectively becomes enforced with Babel 7 since it treats the type annotations as field initializers. Which means that these fields always gets created.

We happened to only use this newer plugin for class fields internally which broke our builds only there.

2b4546d Use `undefined` instead of missing to represent absent field in descriptors. Fixed all the callsites that checks for `hasOwnProperty` or `in` that I could find.

922d40c Fixes a Flow issue in the text printer.

I filed a [follow up issue for ObjectValue](https://github.com/facebook/prepack/issues/2510) since it has the same problem.

### CODE REVIEW

File: src/descriptors.js [hunk 1]
```diff
@@ -124,28 +124,13 @@ export function cloneDescriptor(d: void | PropertyDescriptor): void | PropertyDe
...
-    if (d1.get !== d2.get) return false;
-  }
-  if (d1.hasOwnProperty("set")) {
-    if (!d2.hasOwnProperty("set")) return false;
-    if (d1.set !== d2.set) return false;
-  }
+  if (d1.writable !== d2.writable) return false;
+  if (d1.enumerable !== d2.enumerable) return false;
+  if (d1.configurable !== d2.configurable) return false;
+  if (d1.value !== undefined) {
```
Comment (NTillmann) [inline src/descriptors.js]: There is some asymmetry here, causing equalDescriptors not to be commutative when one `value` is `undefined` and the other one isn't. That issue already existed in the original code. I doubt that it's intentional? Maybe it gets rescued by the other invariant that the existance of `value` and `get`/`set` are mutually exclusive.

Anyway, for clarity, I'd suggest rewriting this case to
```js
if ((d1.value === undefined) !== (d2.value === undefined)) return false;
```

Comment (sebmarkbage) [reply src/descriptors.js]: There are three types of descriptors at the spec level. Generic, Data and Accessors. Lack of value and get/set means it's generic. So if one has a value but the other doesn't, it means that they're not even in the same category of descriptors.

Undefined value doesn't mean `empty` btw. That's an empty value.

This helper function, for reasons that preceded me, doesn't compare whether the values themselves are equivalent.

Basically, this function is only meant to determine if this needs to use joined Descriptor or if it's enough to use a joined Value.

In theory, accessors doesn't need to use a joined Descriptor neither. We can just use a joined value of the getter and a joined value of the setter.

**Comment by trueadm** [comment]
I believe I found the source of the issue. It was to do with the `for (let field in Desc) {` in `properties.js` where we check if the descriptors are identical. Beforehand, we only iterated through properties on the descriptor that existed. Now that they all exist, but are set to undefined this matching logic would fail in many cases and emit undefined properties unnecessarily.
